<!DOCTYPE html><html><head><title>R: Immune-disease associated variants, regions and genes from...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.css">
<script type="text/javascript">
const macros = { "\\R": "\\textsf{R}", "\\code": "\\texttt"};
function processMathHTML() {
    var l = document.getElementsByClassName('reqn');
    for (let e of l) { katex.render(e.textContent, e, { throwOnError: false, macros }); }
    return;
}</script>
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.js"
    onload="processMathHTML();"></script>
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body><div class="container">

<table style="width: 100%;"><tr><td>ImmunoBase</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Immune-disease associated variants, regions and genes from ImmunoBase (hg19)</h2>

<h3>Description</h3>

<p>This dataset contains data obtained from ImmunoBase. For each of 20
immune-diseases, its associated variants, regions, and nearby genes
(within 500kb) are stored.
</p>


<h3>Usage</h3>

<pre><code class='language-R'>data(ImmunoBase)
</code></pre>


<h3>Value</h3>

<p>a list with 5 components:
</p>

<ul>
<li><p><code>disease</code>: a character of disease name
</p>
</li>
<li><p><code>variants</code>: an object of class &quot;GRanges&quot;, storing genomic
locations of associated variants plus their significance and odd
ratios
</p>
</li>
<li><p><code>regions</code>: an object of class &quot;GRanges&quot;, storing genomic
locations of associated regions
</p>
</li>
<li><p><code>genes_variants</code>: a named vector for nearby genes within
500kb of associated variants; the element names are gene symbols, the
element values for the shortest distance to all associated variants
</p>
</li>
<li><p><code>genes_regions</code>: a named vector for nearby genes within
500kb of associated regions; the element names are gene symbols, the
element values for the shortest distance to all associated regions
</p>
</li></ul>



<h3>Examples</h3>

<pre><code class='language-R'>## Not run: 
ImmunoBase &lt;- xRDataLoader(RData.customised='ImmunoBase')
names(ImmunoBase)
ImmunoBase$AS$disease
ImmunoBase$AS$variants
head(ImmunoBase$AS$genes_variants)
head(ImmunoBase$AS$genes_regions)

## End(Not run)
</code></pre>


</div>
</body></html>
