<!DOCTYPE html><html><head><title>R: Table S2A for cis-eQTLs among shared datasets from Benjamin...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.css">
<script type="text/javascript">
const macros = { "\\R": "\\textsf{R}", "\\code": "\\texttt"};
function processMathHTML() {
    var l = document.getElementsByClassName('reqn');
    for (let e of l) { katex.render(e.textContent, e, { throwOnError: false, macros }); }
    return;
}</script>
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.js"
    onload="processMathHTML();"></script>
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body><div class="container">

<table style="width: 100%;"><tr><td>JKscience_TS2A</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Table S2A for cis-eQTLs among shared datasets from Benjamin et al. (2014)</h2>

<h3>Description</h3>

<p>This dataset involves 228 individuals with expression data for all four
conditions, that is, in the naive state (Naive), after 2-hour LPS
(LPS2), after 24-hour LPS (LPS24), and after exposure to IFN (IFN).
Local, likely cis-acting eQTL (referred to as cis-eQTL) were defined as
SNPs showing association with gene expression that were located within
a 1-Mb window of the associated probe. The eQTL analysis was performed
with the R package MatrixEQTL using an additive linear model, reporting
both test statistic and FDR.
</p>


<h3>Usage</h3>

<pre><code class='language-R'>data(JKscience_TS2A)
</code></pre>


<h3>Value</h3>

<p>a data frame. It has the following columns: &quot;variant&quot; (cis-eQTLs),
&quot;ArrayAddress&quot; (illuminaHumanv4), &quot;GeneID&quot; (Entrez GeneID), &quot;Symbol&quot;
(gene symbol), &quot;Naive_t&quot; (test statistic for naive samples), &quot;LPS2_t&quot;,
&quot;LPS24_t&quot;, &quot;IFN_t&quot;, &quot;Naive_fdr&quot; (FDR for naive samples), &quot;LPS2_fdr&quot;,
&quot;LPS24_fdr&quot; and &quot;IFN_fdr&quot;.
</p>


<h3>References</h3>

<p>Fairfax et al. (2014). Innate immune activity conditions the effect of
regulatory variants upon monocyte gene expression. <em>Science</em>,
343(6175):1246949.
</p>


<h3>Examples</h3>

<pre><code class='language-R'>## Not run: 
JKscience_TS2A &lt;- xRDataLoader(RData.customised='JKscience_TS2A')
JKscience_TS2A[1:5,]

## End(Not run)
</code></pre>


</div>
</body></html>
