<!DOCTYPE html><html><head><title>R: Function to obtain separator index.</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.css">
<script type="text/javascript">
const macros = { "\\R": "\\textsf{R}", "\\code": "\\texttt"};
function processMathHTML() {
    var l = document.getElementsByClassName('reqn');
    for (let e of l) { katex.render(e.textContent, e, { throwOnError: false, macros }); }
    return;
}</script>
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.js"
    onload="processMathHTML();"></script>
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body><div class="container">

<table style="width: 100%;"><tr><td>xGRsep</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Function to obtain separator index.</h2>

<h3>Description</h3>

<p><code>xGRsep</code> is supposed to obtain separator index.
</p>


<h3>Usage</h3>

<pre><code class='language-R'>xGRsep(data)
</code></pre>


<h3>Arguments</h3>

<table>
<tr style="vertical-align: top;"><td><code>data</code></td>
<td>
<p>input genomic regions (GR). GR should be provided as a
vector in the format of 'chrN:start-end', where N is either 1-22 or X,
start (or end) is genomic positional number; for example, 'chr1:13-20'</p>
</td></tr>
</table>


<h3>Value</h3>

<p>a vector for separator index
</p>


<h3>See Also</h3>

<p><code>xGRsep</code>
</p>


<h3>Examples</h3>

<pre><code class='language-R'>## Not run: 
# Load the XGR package and specify the location of built-in data
library(XGR)
RData.location &lt;- "http://galahad.well.ox.ac.uk/bigdata"

# a) provide the genomic regions
## load ImmunoBase
ImmunoBase &lt;- xRDataLoader(RData.customised='ImmunoBase',
RData.location=RData.location)
## get lead SNPs reported in AS GWAS and their significance info (p-values)
gr &lt;- ImmunoBase$AS$variant
cse &lt;- xGRcse(gr)

# b) sort index
ind &lt;- xGRsort(cse)
data &lt;- cse[ind]

# c) get separator index
vec_sep &lt;- xGRsep(data)

## End(Not run)
</code></pre>


</div>
</body></html>
